### PR TITLE
Implement proposer's head

### PR DESCRIPTION
**Pull request by potuz**
This PR implements a decoupled `ProposerHead()` from forkchoice's `Head()`. This is the head that proposers will be using to base their blocks. The following logic is applied. 

There are four constants that control the behavior of orphaning late blocks. These constants are specified in this PR with fixed values, but they will be measured statistically for optimal performance before we deploy this mechanism. 
- `OrphanLateBlockFirstThreshold`: we do not orphan blocks before this threshold (default value 4 seconds)
- `PRocessAttestationThreshold`: The first check during the slot to see if we revert our decision to orphan a late block (default value 10 seconds)
- `OrphanLateBlockBalanceFraction`: The maximum percentage of the committee vote that we will allow a block to have and still consider it to being orphaned (default value at 10%)
- `liveFraction`: The minimum percentage of the committee vote that we will require for the previous head in order to apply this mechanism: we will not orphan late blocks if our previous head was unhealthy to not exponentiate liveliness issues on the chain (default value 80%).

The following cases describe the expected behavior if proposers use `ProposerHead()` to base their block
- If the last received block arrived before `OrphanLateBlockFirstThreshold`: the proposer head will equal the canonical head. This applies both in the case that the latest block does become head or not. 
- While we do not see any block in the current slot, we do not change behavior (the proposer head and the current head remain the same). 
- If the currents slot's block arrives late and the current head does not have `liveFraction` votes, the proposer head and the canonical head are equal. 
- If the current slot's block arrives late, and receives more than `OrphanLateBlockBalanceFraction` votes, then the proposer head and the canonical head are equal. 
- If the current canonical head is not a direct child of the previous canonical chain, then the proposer head and the canonical head are equal. 
- If the current slot's block arrives late and does not receive enough votes and the current head has more than `liveFraction` votes and the new head is a direct child of the current head:  this is the only case in which `ProposerHead` will equal the previous head, and `Head()` will equal the canonical head computed without any alterations. 

This PR assumes that blockchain package **will call to process attestations**  and compute head at `ProcessAttestationThreshold` in addition to the beginning of each slot. 

cc. @sproul

### CODE REVIEW

**Review by terencechain — CHANGES_REQUESTED** [review]
Since we dont have codeowner for forkchoice, I'm going to block this PR for review, and giving myself a hard deadline by end of the week

File: beacon-chain/forkchoice/doubly-linked-tree/node.go
Comment (rauljordan) [inline beacon-chain/forkchoice/doubly-linked-tree/node.go:131]: This feels like it definitely belongs in the `slots` package rather than here

File: beacon-chain/forkchoice/doubly-linked-tree/store.go
Comment (rauljordan) [inline beacon-chain/forkchoice/doubly-linked-tree/store.go:16]:
```suggestion
// require before orphaning its child
```

File: beacon-chain/forkchoice/doubly-linked-tree/store.go
Comment (rauljordan) [inline beacon-chain/forkchoice/doubly-linked-tree/store.go:17]: liveFraction is not that informative for a package global. Either make it more descriptive or we can move it as a global config parameter

File: beacon-chain/forkchoice/doubly-linked-tree/store.go
Comment (rauljordan) [inline beacon-chain/forkchoice/doubly-linked-tree/store.go:91]: The complexity of the head function is immense. I know we prefer to have big unit tests for its behavior, but it is becoming harder to deal with when we refactor. There are a lot of tricky edge-cases in here as well that are behind multiple conditionals and mathematical computations I am concerned about adding to an already big function

File: beacon-chain/forkchoice/doubly-linked-tree/store.go
Comment (rauljordan) [inline beacon-chain/forkchoice/doubly-linked-tree/store.go:100]: Duplicated logic with the other helper method that was added to this PR

File: beacon-chain/forkchoice/doubly-linked-tree/store.go
Comment (rauljordan) [inline beacon-chain/forkchoice/doubly-linked-tree/store.go:104]: seems like a useful helper we can abstract `isEarlyBlock()`

File: beacon-chain/forkchoice/doubly-linked-tree/store.go
Comment (rauljordan) [inline beacon-chain/forkchoice/doubly-linked-tree/store.go:113]: Also ideal to compute the votedFraction as an independent, pure function to reduce complexity here

File: beacon-chain/forkchoice/doubly-linked-tree/store.go
Comment (rauljordan) [inline beacon-chain/forkchoice/doubly-linked-tree/store.go:157]: Also helper function for is future block here

File: beacon-chain/forkchoice/doubly-linked-tree/store.go
Comment (rauljordan) [inline beacon-chain/forkchoice/doubly-linked-tree/store.go:189]: duplicated logic

File: config/params/mainnet_config.go
Comment (rauljordan) [inline config/params/mainnet_config.go:191]: We're already adding a few new globals, might as well add the required fraction of committee votes for orphaning late blocks here

**Comment by terencechain** [comment]
> The way the proposer head is computed in the `head()` function itself is a side-effect. Instead, if it can be computed on the fly by an actual proposer when needed, it would reduce complexity of the head function and make it easier to write code that reads like the design doc for this pull request. Ideally, something like this, where `X` and `Y` are defined such as in the document https://www.notion.so/prysmaticlabs/Enabling-proposer-to-reorg-against-late-parent-block-55761f83c72442dd98d6db033122dfa3
> 
> ```
> func APIGetBlock() {
>    parentRoot := forkchoice.ComputeProposerHead()
> }
> 
> func computeProposerHead() {
> if seenBlockBeforeX() {
>   return normalLogic()
> }
> if inX {
>   return xIntervalLogic()
> }
> return yIntervalLogic()
> }
> ```
> 
> and each interval's logic for deciding the block the proposer will build on reads cleanly

Here is the blockchain changes for reference: https://github.com/prysmaticlabs/prysm/pull/11426

This PR only encapsulates all the forkchoice chances. With that said, I still think your feedback is really good

File: beacon-chain/forkchoice/doubly-linked-tree/node.go
Comment (potuz) [reply beacon-chain/forkchoice/doubly-linked-tree/node.go:131]: There is such a helper function in the `slots` package, that takes a time. instead of calling this as a method in the node I could have the caller using that helper function and passing it the genesis time and the node's timestamp. I wanted to avoid the parameter passing, but if you think it's better to use that method I can do that.

File: beacon-chain/forkchoice/doubly-linked-tree/store.go
Comment (potuz) [reply beacon-chain/forkchoice/doubly-linked-tree/store.go:189]: I didn't understand this one

File: config/params/mainnet_config.go
Comment (potuz) [reply config/params/mainnet_config.go:191]: I was worried about this, doing so requires calling a function from the forkchoice package to get back the global parameters and then request the constant. This constant can only be useful from within forkchoice. Is it really worth it to have it as a beacon-chain parameter?

File: beacon-chain/forkchoice/doubly-linked-tree/store.go
Comment (potuz) [reply beacon-chain/forkchoice/doubly-linked-tree/store.go:91]: > The way the proposer head is computed in the `head()` function itself is a side-effect. Instead, if it can be computed on the fly by an actual proposer when needed, it would reduce complexity of the head function and make it easier to write code that reads like the design doc for this pull request. Ideally, something like this, where `X` and `Y` are defined such as in the document https://www.notion.so/prysmaticlabs/Enabling-proposer-to-reorg-against-late-parent-block-55761f83c72442dd98d6db033122dfa3
> 
> ```
> func APIGetBlock() {
>    parentRoot := forkchoice.ComputeProposerHead()
> }
> 
> func computeProposerHead() {
> if seenBlockBeforeX() {
>   return normalLogic()
> }
> if inX {
>   return xIntervalLogic()
> }
> return yIntervalLogic()
> }
> ```
> 
> and each interval's logic for deciding the block the proposer will build on reads cleanly

This can be achieved only by increasing the memory footprint of forkchoice and I'm not convinced it will reduce the complexity. The only way I can now think how to implement this this way would be by keeping cached the number of votes that the late block had at `y`. I'll give it a shot and we can compare both approaches.

File: beacon-chain/forkchoice/doubly-linked-tree/store.go
Comment (rauljordan) [reply beacon-chain/forkchoice/doubly-linked-tree/store.go:189]: ah i mean that i've seen a few lines of code in the PR that compute seconds into a slot that could just a single helper function

File: config/params/mainnet_config.go
Comment (rauljordan) [reply config/params/mainnet_config.go:191]: Fair enough, perhaps we can just give it a more descriptive name to know that it is the safe number of committee votes to enable proposer reorgs

File: beacon-chain/forkchoice/doubly-linked-tree/store.go
Comment (rauljordan) [reply beacon-chain/forkchoice/doubly-linked-tree/store.go:91]: Understood - I do think we should still simplify some of this logic with cleaner helper functions as it is hard to reason about in isolation. Specifically, the computations we use to decide how to set the proposer head

**Comment by prestonvanloon** [comment]
@potuz what's the status on this PR?

**Comment by potuz** [comment]
> @potuz what's the status on this PR?

Need to review it again due to changes in our design. Waiting for the corresponding spec PR to be merged, will move this to draft in the meantime. Thanks for the ping

**Comment by potuz** [comment]
Closing in favor of #11981

**Pull request closed without merging**
